src/wp-admin/includes/class-rollback-auto-update.php

```diff
@@ -0,0 +1,551 @@
+		$this->handler_args['handler_error'] = 'RAU Exception Handler';
+		$this->handler_args['error_msg']     = $exception->getMessage();
+		$this->handler();
+	}
+
+	/**
+	 * Shutdown function.
+	 *
+	 * @return array|void
+	 */
+	public function shutdown_function() {
+		$last_error = error_get_last();
+		$result     = $this->check_passing_errors( $last_error['message'] );
+		if ( is_array( $result ) ) {
+			$this->restart_updates_and_send_email();
+			exit();
+		}
+		$this->handler_args['handler_error'] = 'RAU Shutdown Function';
+		$this->handler_args['error_msg']     = $last_error['message'];
+		$this->handler();
+	}
+
+	/**
+	 * Check for errors only caused by an active plugin using 'include()'.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param string $error_msg Error message from handler.
+	 * @return array|bool
+	 */
+	private function check_passing_errors( $error_msg ) {
+		if ( empty( $error_msg ) ) {
+			return array();
+		}
+		preg_match( '/(' . implode( '|', static::$error_exceptions ) . ')/', $error_msg, $matches );
+		if ( ! empty( $matches ) ) {
+			return $this->handler_args['result'];
+		}
+
+		return false;
+	}
+
+	/**
+	 * Handles errors by running Rollback.
+	 *
+	 * @since 6.4.0
+	 */
+	private function handler() {
+		error_log( var_export( $this->handler_args['handler_error'] . ' - ' . $this->handler_args['error_msg'], true ) );
+		if ( in_array( $this->handler_args['hook_extra']['plugin'], self::$fatals, true ) ) {
+			return;
+		}
+		self::$fatals[] = $this->handler_args['hook_extra']['plugin'];
+
+		$this->cron_rollback();
+
+		/*
+		 * This possibly helps to avoid a potential race condition on servers that may start to
+		 * process the next plugin for auto-updating before the handler can pick up an error from
+		 * the previously processed plugin.
+		 */
+		sleep( 2 );
+
+		/*
+		 * If a plugin upgrade fails prior to a theme upgrade running, the plugin upgrader will have
+		 * hooked the 'Plugin_Upgrader::delete_old_plugin()' method to 'upgrader_clear_destination',
+		 * which will return a `WP_Error` object and prevent the process from continuing.
+		 *
+		 * To resolve this, the hook must be removed using the original plugin upgrader instance.
+		 */
+		remove_filter( 'upgrader_clear_destination', array( static::$plugin_upgrader, 'delete_old_plugin' ) );
+
+		$this->restart_updates_and_send_email();
+	}
+
+	/**
+	 * Rolls back during cron.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @global WP_Filesystem_Base $wp_filesystem WordPress filesystem subclass.
+	 */
+	private function cron_rollback() {
+		global $wp_filesystem;
+
+		$temp_backup = array(
+			'temp_backup' => array(
+				'dir'  => 'plugins',
+				'slug' => dirname( $this->handler_args['hook_extra']['plugin'] ),
+				'src'  => $wp_filesystem->wp_plugins_dir(),
+			),
+		);
+
+		$rollback_updater = new WP_Upgrader();
+
+		// Set private $temp_restores variable.
+		$ref_temp_restores = new ReflectionProperty( $rollback_updater, 'temp_restores' );
```

Comment:
It's not the first use of Reflection in core.

https://github.com/WordPress/wordpress-develop/blob/1e11c3fce87f00520d2a46322c1dc8edc9952dd7/src/wp-admin/includes/template.php#L1248-L1259

As the comment states, it's setting a private variable. I'm not sure I understand, unless you're just looking for a more verbose comment explanation. If that's the case, no problem.

The only way to avoid the Reflection is to make all the variables and methods referenced in core in the public scope.

Reply:
@getsource Inline comments adding context have now been added where the Reflection API is used in the class.

Reply:
This looks great, thank you!

And thanks @afragen -- I had not found that one in my search.